<h1 align=center><font size = 5>Segmenting and Clustering Neighborhoods in Toronto</font></h1>

Import libraries

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


<a id='item1'></a>

## 1. Download and Explore Dataset

In [2]:
!wget -q -O 'geospatial_toronto.csv' https://cocl.us/Geospatial_data
print('Data downloaded!')

Data downloaded!


#### Load and explore the data

Next, let's load the data.

In [3]:
neigh = pd.read_csv('geospatial_toronto.csv')
neigh.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

Let's take a quick look at the data.

In [4]:
neigh.info
neigh.shape

(103, 3)

In [5]:
neigh1 = pd.read_csv('Toronto_neigh.csv')
neigh1.head()

Unnamed: 0 PostalCode       Borough  \
0           0        M1B   Scarborough   
1           1        M1C   Scarborough   
2           2        M1E   Scarborough   
3           3        M1G   Scarborough   
4           4        M1H   Scarborough   

                              Neighborhood  
0                           Rouge, Malvern  
1   Highland Creek, Rouge Hill, Port Union  
2        Guildwood, Morningside, West Hill  
3                                   Woburn  
4                                Cedarbrae

In [6]:
neigh1.info
neigh1.shape

(103, 4)

In [7]:
neighborhoods = neigh1.merge ( neigh, left_on='PostalCode', right_on='Postal Code', how='inner')[['Postal Code', 'Borough', 'Neighborhood', 'Latitude', 'Longitude']]

In [8]:
neighborhoods.info
neighborhoods.shape

(103, 5)

In [9]:
neighborhoods.head(10)

Postal Code       Borough                                      Neighborhood  \
0         M1B   Scarborough                                    Rouge, Malvern   
1         M1C   Scarborough            Highland Creek, Rouge Hill, Port Union   
2         M1E   Scarborough                 Guildwood, Morningside, West Hill   
3         M1G   Scarborough                                            Woburn   
4         M1H   Scarborough                                         Cedarbrae   
5         M1J   Scarborough                               Scarborough Village   
6         M1K   Scarborough       East Birchmount Park, Ionview, Kennedy Park   
7         M1L   Scarborough                   Clairlea, Golden Mile, Oakridge   
8         M1M   Scarborough   Cliffcrest, Cliffside, Scarborough Village West   
9         M1N   Scarborough                       Birch Cliff, Cliffside West   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476  
5  43.744734 -79.239476  
6  43.727929 -79.262029  
7  43.711112 -79.284577  
8  43.716316 -79.239476  
9  43.692657 -79.264848

And make sure that the dataset has all 11 boroughs and 103 neighborhoods.

In [10]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape[0]
    )
)

The dataframe has 11 boroughs and 103 neighborhoods.


In [11]:
address = 'Toronto, CA'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


#### Create a map of TORONTO with neighborhoods superimposed on top.

In [12]:
# create map of New York using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

**Folium** is a great visualization library. Feel free to zoom into the above map, and click on each circle mark to reveal the name of the neighborhood and its respective borough.

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

#### Define Foursquare Credentials and Version

In [13]:
CLIENT_ID = 'GAUZTZGPUWZ2PFHAZTVJTVELJU2UOYIDFHOMUFIXJLLN2YWX' # your Foursquare ID
CLIENT_SECRET = 'E3RW5IVI1MZ33YZBLZOROE3REFRT0GEOKH3HCUKWHEZRWUSO' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: GAUZTZGPUWZ2PFHAZTVJTVELJU2UOYIDFHOMUFIXJLLN2YWX
CLIENT_SECRET:E3RW5IVI1MZ33YZBLZOROE3REFRT0GEOKH3HCUKWHEZRWUSO


#### Let's explore the first neighborhood in our dataframe.

Get the neighborhood's name.

In [14]:
neighborhoods.loc[0, 'Neighborhood']

' Rouge, Malvern'

Get the neighborhood's latitude and longitude values.

In [15]:
neighborhood_latitude = neighborhoods.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = neighborhoods.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = neighborhoods.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of  Rouge, Malvern are 43.806686299999996, -79.19435340000001.


#### Now, let's get the top 100 venues that are in ROUGE MALVERN within a radius of 500 meters.

First, let's create the GET request URL. Name your URL **url**.

In [16]:
# type your answer here

LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius
# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=GAUZTZGPUWZ2PFHAZTVJTVELJU2UOYIDFHOMUFIXJLLN2YWX&client_secret=E3RW5IVI1MZ33YZBLZOROE3REFRT0GEOKH3HCUKWHEZRWUSO&v=20180605&ll=43.806686299999996,-79.19435340000001&radius=500&limit=100'

Double-click __here__ for the solution.
<!-- The correct answer is:
LIMIT = 100 # limit of number of venues returned by Foursquare API
-->

<!--
radius = 500 # define radius
-->

<!--
\\ # create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL
--> 

Send the GET request and examine the resutls

In [17]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5c6befb94c1f676352f811bb'},
 'response': {'warning': {'text': "There aren't a lot of results near you. Try something more general, reset your filters, or expand the search area."},
  'headerLocation': 'Malvern',
  'headerFullLocation': 'Malvern, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 1,
  'suggestedBounds': {'ne': {'lat': 43.8111863045, 'lng': -79.18812958073042},
   'sw': {'lat': 43.80218629549999, 'lng': -79.2005772192696}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4bb6b9446edc76b0d771311c',
       'name': "Wendy's",
       'location': {'crossStreet': 'Morningside & Sheppard',
        'lat': 43.80744841934756,
        'lng': -79.19905558052072,
        'labeledLatLngs': [{'label': 'display'

From the Foursquare lab in the previous module, we know that all the information is in the *items* key. Before we proceed, let's borrow the **get_category_type** function from the Foursquare lab.

In [18]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a *pandas* dataframe.

In [19]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name            categories        lat        lng
0  Wendy's  Fast Food Restaurant  43.807448 -79.199056

And how many venues were returned by Foursquare?

In [20]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

1 venues were returned by Foursquare.


<a id='item2'></a>

## 2. Explore Neighborhoods in TORONTO

#### Let's create a function to repeat the same process to all the neighborhoods in Toronto

In [21]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now write the code to run the above function on each neighborhood and create a new dataframe called *toronto_venues*.

In [22]:
# type your answer here

toronto_venues = getNearbyVenues(names=neighborhoods['Neighborhood'],
                                   latitudes=neighborhoods['Latitude'],
                                   longitudes=neighborhoods['Longitude']
                                  )



 Rouge, Malvern
 Highland Creek, Rouge Hill, Port Union
 Guildwood, Morningside, West Hill
 Woburn
 Cedarbrae
 Scarborough Village
 East Birchmount Park, Ionview, Kennedy Park
 Clairlea, Golden Mile, Oakridge
 Cliffcrest, Cliffside, Scarborough Village West
 Birch Cliff, Cliffside West
 Dorset Park, Scarborough Town Centre, Wexford Heights
 Maryvale, Wexford
 Agincourt
 Clarks Corners, Sullivan, Tam O'Shanter
 Agincourt North, L'Amoreaux East, Milliken, Steeles East
 L'Amoreaux West, Steeles West
 Upper Rouge
 Hillcrest Village
 Fairview, Henry Farm, Oriole
 Bayview Village
 Silver Hills, York Mills
 Newtonbrook, Willowdale
 Willowdale South
 York Mills West
 Willowdale West
 Parkwoods
 Don Mills North
 Flemingdon Park, Don Mills South
 Bathurst Manor, Downsview North, Wilson Heights
 Northwood Park, York University
 CFB Toronto, Downsview East
 Downsview West
 Downsview Central
 Downsview Northwest
 Victoria Village
 Woodbine Gardens, Parkview Hill
 Woodbine Heights
 The Beaches
 Leas

Double-click __here__ for the solution.
<!-- The correct answer is:
manhattan_venues = getNearbyVenues(names=manhattan_data['Neighborhood'],
                                   latitudes=manhattan_data['Latitude'],
                                   longitudes=manhattan_data['Longitude']
                                  )
--> 

#### Let's check the size of the resulting dataframe

In [23]:
print(toronto_venues.shape)
toronto_venues.head()

(2236, 7)


Neighborhood  Neighborhood Latitude  \
0                           Rouge, Malvern              43.806686   
1   Highland Creek, Rouge Hill, Port Union              43.784535   
2   Highland Creek, Rouge Hill, Port Union              43.784535   
3        Guildwood, Morningside, West Hill              43.763573   
4        Guildwood, Morningside, West Hill              43.763573   

   Neighborhood Longitude                            Venue  Venue Latitude  \
0              -79.194353                          Wendy's       43.807448   
1              -79.160497            Royal Canadian Legion       43.782533   
2              -79.160497        Affordable Toronto Movers       43.787919   
3              -79.188711  Swiss Chalet Rotisserie & Grill       43.767697   
4              -79.188711                G & G Electronics       43.765309   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.163085                   Bar  
2       -79.162977         Moving Target  
3       -79.189914           Pizza Place  
4       -79.191537     Electronics Store

Let's check how many venues were returned for each neighborhood

In [24]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
 Adelaide, King, Richmond                                             100   
 Agincourt                                                              5   
 Agincourt North, L'Amoreaux East, Milliken, St...                      2   
 Albion Gardens, Beaumond Heights, Humbergate, ...                     10   
 Alderwood, Long Branch                                                 8   
 Bathurst Manor, Downsview North, Wilson Heights                       16   
 Bayview Village                                                        4   
 Bedford Park, Lawrence Manor East                                     25   
 Berczy Park                                                           55   
 Birch Cliff, Cliffside West                                            4   
 Bloordale Gardens, Eringate, Markland Wood, Ol...                      7   
 Brockton, Exhibition Place, Parkdale Village                          19   
 Business Reply Mail Processing Centre 969 Eastern                     19   
 CFB Toronto, Downsview East                                            3   
 CN Tower, Bathurst Quay, Island airport, Harbo...                     14   
 Cabbagetown, St. James Town                                           44   
 Caledonia-Fairbanks                                                    6   
 Canada Post Gateway Processing Centre                                 11   
 Cedarbrae                                                              7   
 Central Bay Street                                                    81   
 Chinatown, Grange Park, Kensington Market                            100   
 Christie                                                              15   
 Church and Wellesley                                                  87   
 Clairlea, Golden Mile, Oakridge                                       10   
 Clarks Corners, Sullivan, Tam O'Shanter                                9   
 Cliffcrest, Cliffside, Scarborough Village West                        2   
 Cloverdale, Islington, Martin Grove, Princess ...                      1   
 Commerce Court, Victoria Hotel                                       100   
 Davisville                                                            32   
 Davisville North                                                       8   
 Deer Park, Forest Hill SE, Rathnelly, South Hi...                     15   
 Del Ray, Keelesdale, Mount Dennis, Silverthorn                         4   
 Design Exchange, Toronto Dominion Centre                             100   
 Don Mills North                                                        6   
 Dorset Park, Scarborough Town Centre, Wexford ...                      7   
 Dovercourt Village, Dufferin                                          20   
 Downsview Central                                                      4   
 Downsview Northwest                                                    4   
 Downsview West                                                         4   
 East Birchmount Park, Ionview, Kennedy Park                            5   
 East Toronto                                                           4   
 Emery, Humberlea                                                       1   
 Fairview, Henry Farm, Oriole                                          64   
 First Canadian Place, Underground city                               100   
 Flemingdon Park, Don Mills South                                      21   
 Forest Hill North, Forest Hill West                                    4   
 Glencairn                                                              4   
 Guildwood, Morningside, West Hill                                      6   
 Harbord, University of Toronto                                        36   
 Harbourfront East, Toronto Islands, Union Station                    100   
 Harbourfront, Regent Park                      

#### Let's find out how many unique categories can be curated from all the returned venues

In [25]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 276 uniques categories.


<a id='item3'></a>

## 3. Analyze Each Neighborhood

In [26]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Accessories Store  Adult Boutique  Afghan Restaurant  Airport  \
0            0                  0               0                  0        0   
1            0                  0               0                  0        0   
2            0                  0               0                  0        0   
3            0                  0               0                  0        0   
4            0                  0               0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  \
0                 0                    0             0         0   
1                 0                    0             0         0   
2                 0                    0             0         0   
3                 0                    0             0         0   
4                 0                    0             0         0   

   Arepa Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0                 0            0           0                    0   
1                 0            0           0                    0   
2                 0            0           0                    0   
3                 0            0           0                    0   
4                 0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Garage  Auto Workshop  \
0                 0                   0            0              0   
1                 0                   0            0              0   
2                 0                   0            0              0   
3                 0                   0            0              0   
4                 0                   0            0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0          0           0           0       0     0    0               0   
1          0           0           0       0     0    1               0   
2          0           0           0       0     0    0               0   
3          0           0           0       0     0    0               0   
4          0           0           0       0     0    0               0   

   Baseball Stadium  Basketball Court  Basketball Stadium  Beach  Beer Bar  \
0                 0                 0                   0      0         0   
1                 0                 0                   0      0         0   
2                 0                 0                   0      0         0   
3                 0                 0                   0      0         0   
4                 0                 0                   0      0         0   

   Beer Store  Belgian Restaurant  Bike Shop  Bistro  Boat or Ferry  \
0           0                   0          0       0              0   
1           0                   0          0       0              0   
2           0                   0          0       0              0   
3           0                   0          0       0              0   
4           0                   0          0       0              0   

   Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  \
0          0         0                     0               0        0   
1          0         0                     0               0        0   
2          0         0                     0               0        0   
3          0         0                     0               0        0   
4          0         0                     0               0        0   

   Bridal Shop  Bubble Tea Shop  Building  Bur

And let's examine the new dataframe size.

In [27]:
toronto_onehot.shape

(2236, 276)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [28]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                            Adelaide, King, Richmond     0.000000   
1                                           Agincourt     0.000000   
2    Agincourt North, L'Amoreaux East, Milliken, S...     0.000000   
3    Albion Gardens, Beaumond Heights, Humbergate,...     0.000000   
4                              Alderwood, Long Branch     0.000000   
5     Bathurst Manor, Downsview North, Wilson Heights     0.000000   
6                                     Bayview Village     0.000000   
7                   Bedford Park, Lawrence Manor East     0.000000   
8                                         Berczy Park     0.000000   
9                         Birch Cliff, Cliffside West     0.000000   
10   Bloordale Gardens, Eringate, Markland Wood, O...     0.000000   
11       Brockton, Exhibition Place, Parkdale Village     0.052632   
12   Business Reply Mail Processing Centre 969 Eas...     0.000000   
13                        CFB Toronto, Downsview East     0.000000   
14   CN Tower, Bathurst Quay, Island airport, Harb...     0.000000   
15                        Cabbagetown, St. James Town     0.000000   
16                                Caledonia-Fairbanks     0.000000   
17              Canada Post Gateway Processing Centre     0.000000   
18                                          Cedarbrae     0.000000   
19                                 Central Bay Street     0.012346   
20          Chinatown, Grange Park, Kensington Market     0.000000   
21                                           Christie     0.000000   
22                               Church and Wellesley     0.011494   
23                    Clairlea, Golden Mile, Oakridge     0.000000   
24            Clarks Corners, Sullivan, Tam O'Shanter     0.000000   
25    Cliffcrest, Cliffside, Scarborough Village West     0.000000   
26   Cloverdale, Islington, Martin Grove, Princess...     0.000000   
27                     Commerce Court, Victoria Hotel     0.000000   
28                                         Davisville     0.000000   
29                                   Davisville North     0.000000   
30   Deer Park, Forest Hill SE, Rathnelly, South H...     0.000000   
31     Del Ray, Keelesdale, Mount Dennis, Silverthorn     0.000000   
32           Design Exchange, Toronto Dominion Centre     0.000000   
33                                    Don Mills North     0.000000   
34   Dorset Park, Scarborough Town Centre, Wexford...     0.000000   
35                       Dovercourt Village, Dufferin     0.000000   
36                                  Downsview Central     0.000000   
37                                Downsview Northwest     0.000000   
38                                     Downsview West     0.000000   
39        East Birchmount Park, Ionview, Kennedy Park     0.000000   
40                                       East Toronto     0.000000   
41                                   Emery, Humberlea     0.000000   
42                       Fairview, Henry Farm, Oriole     0.000000   
43             First Canadian Place, Underground city     0.000000   
44                   Flemingdon Park, Don Mills South     0.000000   
45                Forest Hill North, Forest Hill West     0.000000   
46                                          Glencairn     0.000000   
47                  Guildwood, Morningside, West Hill     0.000000   
48                     Harbord, University of Toronto     0.027778   
49   Harbourfront East, Toronto Islands, Union Sta...     0.000000   
50                          Harbourfront, Regent Park     0.000000   
51                      High Park, The Junction South     0.000000   
52             Highland Creek, Rouge Hill, Port Union     0.000000   
53                                  Hillcrest Village     0.000000   
54       Humber Bay Shores, Mimico South, New Toronto     0.000000   
55   Humber Bay, King's Mill Park, Kingsway Park S...     0.000000   
56                                      Humber Summ

#### Let's confirm the new size

In [29]:
toronto_grouped.shape

(99, 276)

#### Let's print each neighborhood along with the top 5 most common venues

In [30]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

---- Adelaide, King, Richmond----
                 venue  freq
0          Coffee Shop  0.06
1                 Café  0.05
2           Steakhouse  0.04
3  American Restaurant  0.04
4      Thai Restaurant  0.04


---- Agincourt----
                venue  freq
0      Sandwich Place   0.2
1      Breakfast Spot   0.2
2              Lounge   0.2
3        Skating Rink   0.2
4  Chinese Restaurant   0.2


---- Agincourt North, L'Amoreaux East, Milliken, Steeles East----
                       venue  freq
0                 Playground   0.5
1                       Park   0.5
2                Yoga Studio   0.0
3  Middle Eastern Restaurant   0.0
4                      Motel   0.0


---- Albion Gardens, Beaumond Heights, Humbergate, Jamestown, Mount Olive, Silverstone, South Steeles, Thistletown----
                 venue  freq
0        Grocery Store   0.2
1          Coffee Shop   0.1
2       Sandwich Place   0.1
3  Fried Chicken Joint   0.1
4         Liquor Store   0.1


---- Alderwood, Long Branch-

#### Let's put that into a *pandas* dataframe

First, let's write a function to sort the venues in descending order.

In [31]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [32]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                          Agincourt                Lounge   
2   Agincourt North, L'Amoreaux East, Milliken, S...            Playground   
3   Albion Gardens, Beaumond Heights, Humbergate,...         Grocery Store   
4                             Alderwood, Long Branch           Pizza Place   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café   American Restaurant       Thai Restaurant   
1          Skating Rink        Sandwich Place        Breakfast Spot   
2                  Park         Women's Store   Dumpling Restaurant   
3              Pharmacy  Fast Food Restaurant           Pizza Place   
4           Coffee Shop                   Gym              Pharmacy   

  5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0            Steakhouse      Asian Restaurant                          Bar   
1    Chinese Restaurant         Women's Store  Eastern European Restaurant   
2                 Diner        Discount Store                      Dog Run   
3           Coffee Shop            Beer Store                 Liquor Store   
4          Skating Rink        Sandwich Place                          Pub   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                Bakery        Clothing Store             Restaurant  
1      Doner Restaurant            Donut Shop              Drugstore  
2      Doner Restaurant            Donut Shop              Drugstore  
3   Fried Chicken Joint        Sandwich Place             Donut Shop  
4      Doner Restaurant    Dim Sum Restaurant                  Diner

<a id='item4'></a>

## 4. Cluster Neighborhoods

Run *k*-means to cluster the neighborhood into 5 clusters.

In [33]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 1, 0, 0, 0, 3, 0, 0, 0], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [34]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = neighborhoods

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged # check the last columns!

Postal Code            Borough  \
0           M1B        Scarborough   
1           M1C        Scarborough   
2           M1E        Scarborough   
3           M1G        Scarborough   
4           M1H        Scarborough   
5           M1J        Scarborough   
6           M1K        Scarborough   
7           M1L        Scarborough   
8           M1M        Scarborough   
9           M1N        Scarborough   
10          M1P        Scarborough   
11          M1R        Scarborough   
12          M1S        Scarborough   
13          M1T        Scarborough   
14          M1V        Scarborough   
15          M1W        Scarborough   
16          M1X        Scarborough   
17          M2H         North York   
18          M2J         North York   
19          M2K         North York   
20          M2L         North York   
21          M2M         North York   
22          M2N         North York   
23          M2P         North York   
24          M2R         North York   
25          M3A         North York   
26          M3B         North York   
27          M3C         North York   
28          M3H         North York   
29          M3J         North York   
30          M3K         North York   
31          M3L         North York   
32          M3M         North York   
33          M3N         North York   
34          M4A         North York   
35          M4B          East York   
36          M4C          East York   
37          M4E       East Toronto   
38          M4G          East York   
39          M4H          East York   
40          M4J          East York   
41          M4K       East Toronto   
42          M4L       East Toronto   
43          M4M       East Toronto   
44          M4N    Central Toronto   
45          M4P    Central Toronto   
46          M4R    Central Toronto   
47          M4S    Central Toronto   
48          M4T    Central Toronto   
49          M4V    Central Toronto   
50          M4W   Downtown Toronto   
51          M4X   Downtown Toronto   
52          M4Y   Downtown Toronto   
53          M5A   Downtown Toronto   
54          M5B   Downtown Toronto   
55          M5C   Downtown Toronto   
56          M5E   Downtown Toronto   
57          M5G   Downtown Toronto   
58          M5H   Downtown Toronto   
59          M5J   Downtown Toronto   
60          M5K   Downtown Toronto   
61          M5L   Downtown Toronto   
62          M5M         North York   
63          M5N    Central Toronto   
64          M5P    Central Toronto   
65          M5R    Central Toronto   
66          M5S   Downtown Toronto   
67          M5T   Downtown Toronto   
68          M5V   Downtown Toronto   
69          M5W   Downtown Toronto   
70          M5X   Downtown Toronto   
71          M6A         North York   
72          M6B         North York   
73          M6C               York   
74          M6E               York   
75          M6G   Downtown Toronto   
76          M6H       West Toronto   
77          M6J       West Toronto   
78          M6K       West Toronto   
79          M6L         North York   
80          M6M               York   
81          M6N               York   
82          M6P       West Toronto   
83          M6R       West Toronto   
84          M6S       West Toronto   
85          M7A       Queen's Park   
86          M7R        Mississauga   
87          M7Y       East Toronto   
88          M8V          Etobicoke   
89          M8W          Etobicoke   
90          M8X          Etobicoke   
91          M8Y          Etobicoke   
92          M8Z          Etobicoke   
93          M9A          Etobicoke   
94          M9B          Etobicoke   
95          M9C          Etobicoke   
96          M9L         North York   
97          M9M         North York   
98          M9N               York   
99          M9P          Etobicoke   
100         M9R          Etobicoke   
101         M9V          Etobicoke   
102         M9W          Etobicoke   

                                          Neighborh

In [35]:
toronto_merged = toronto_merged.drop ([16,20,21,93])

Finally, let's visualize the resulting clusters

In [36]:
#toronto_merged [['Cluster Labels' == '']]
toronto_merged['Cluster Labels'].isnull().sum().sum()


0

In [37]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    #cluster = int ( cluster )
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int (cluster - 1)],
        fill=True,
        fill_color=rainbow[ int (cluster - 1)],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

<a id='item5'></a>

## 5. Examine Clusters

Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster. I will leave this exercise to you.

#### Cluster 1

In [38]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels      1st Most Common Venue  \
0          Scarborough             0.0       Fast Food Restaurant   
1          Scarborough             0.0              Moving Target   
2          Scarborough             0.0         Mexican Restaurant   
3          Scarborough             0.0                Coffee Shop   
4          Scarborough             0.0         Athletics & Sports   
6          Scarborough             0.0           Department Store   
7          Scarborough             0.0                     Bakery   
8          Scarborough             0.0                      Motel   
9          Scarborough             0.0            College Stadium   
10         Scarborough             0.0          Indian Restaurant   
11         Scarborough             0.0  Middle Eastern Restaurant   
12         Scarborough             0.0                     Lounge   
13         Scarborough             0.0                Pizza Place   
15         Scarborough             0.0       Fast Food Restaurant   
17          North York             0.0       Fast Food Restaurant   
18          North York             0.0             Clothing Store   
22          North York             0.0                 Restaurant   
24          North York             0.0                   Pharmacy   
26          North York             0.0                       Café   
27          North York             0.0                 Beer Store   
28          North York             0.0                Coffee Shop   
29          North York             0.0                        Bar   
31          North York             0.0                      Hotel   
32          North York             0.0           Business Service   
33          North York             0.0             Discount Store   
34          North York             0.0                Coffee Shop   
35           East York             0.0       Fast Food Restaurant   
36           East York             0.0                   Pharmacy   
37        East Toronto             0.0                Coffee Shop   
38           East York             0.0                Coffee Shop   
39           East York             0.0          Indian Restaurant   
41        East Toronto             0.0           Greek Restaurant   
42        East Toronto             0.0             Sandwich Place   
43        East Toronto             0.0                       Café   
45     Central Toronto             0.0                       Park   
46     Central Toronto             0.0             Clothing Store   
47     Central Toronto             0.0             Sandwich Place   
48     Central Toronto             0.0                      Trail   
49     Central Toronto             0.0          Convenience Store   
51    Downtown Toronto             0.0                Coffee Shop   
52    Downtown Toronto             0.0                Coffee Shop   
53    Downtown Toronto             0.0                Coffee Shop   
54    Downtown Toronto             0.0                Coffee Shop   
55    Downtown Toronto             0.0                Coffee Shop   
56    Downtown Toronto             0.0                Coffee Shop   
57    Downtown Toronto             0.0                Coffee Shop   
58    Downtown Toronto             0.0                Coffee Shop   
59    Downtown Toronto             0.0                Coffee Shop   
60    Downtown Toronto             0.0                Coffee Shop   
61    Downtown Toronto             0.0                Coffee Shop   
62          North York             0.0                Coffee Shop   
63     Central Toronto             0.0                     Garden   
64     Central Toronto             0.0                      Trail   
65     Central Toronto             0.0                       Café   
66    Downtown Toronto             0.0                       Café   
67    Downtown Toronto             0.0                       Café   
68    Downtown Toronto             0.0             Airport Lounge   
69    Downtow

#### Cluster 2

In [39]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
14         Scarborough             1.0            Playground   
23          North York             1.0                   Bar   
25          North York             1.0  Fast Food Restaurant   
30          North York             1.0              Bus Stop   
40           East York             1.0                  Park   
44     Central Toronto             1.0              Bus Line   
50    Downtown Toronto             1.0                  Park   
72          North York             1.0                   Pub   
74                York             1.0                  Park   
79          North York             1.0                  Park   
90           Etobicoke             1.0                 River   
98                York             1.0                  Park   
100          Etobicoke             1.0           Pizza Place   

          2nd Most Common Venue 3rd Most Common Venue  \
14                         Park         Women's Store   
23                         Park                  Bank   
25                         Park             BBQ Joint   
30                         Park               Airport   
40                Metro Station     Convenience Store   
44                         Park    Dim Sum Restaurant   
50                        Trail            Playground   
72                  Pizza Place                  Park   
74                Women's Store  Fast Food Restaurant   
79   Construction & Landscaping      Basketball Court   
90                         Park         Women's Store   
98            Convenience Store         Women's Store   
100                        Park              Bus Line   

           4th Most Common Venue        5th Most Common Venue  \
14           Dumpling Restaurant                        Diner   
23                 Women's Store             Doner Restaurant   
25             Food & Drink Shop  Eastern European Restaurant   
30   Eastern European Restaurant               Discount Store   
40                 Women's Store               Discount Store   
44                   Swim School                         Lake   
50           Dumpling Restaurant                        Diner   
72           Japanese Restaurant                Women's Store   
74                        Market                     Pharmacy   
79                        Bakery                Women's Store   
90                     Drugstore                        Diner   
98   Eastern European Restaurant               Discount Store   
100            Mobile Phone Shop                Women's Store   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
14         Discount Store               Dog Run      Doner Restaurant   
23             Donut Shop             Drugstore   Dumpling Restaurant   
25                Dog Run      Doner Restaurant            Donut Shop   
30                Dog Run      Doner Restaurant            Donut Shop   
40                Dog Run      Doner Restaurant            Donut Shop   
44    Dumpling Restaurant               Dog Run      Doner Restaurant   
50         Discount Store               Dog Run      Doner Restaurant   
72                  Diner        Discount Store               Dog Run   
74                Airport                 Diner        Farmers Market   
79      Electronics Store      Doner Restaurant            Donut Shop   
90         Discount Store               Dog Run      Doner Restaurant   
98                Dog Run      Doner Restaurant            Donut Shop   
100   Dumpling Restaurant               Dog Run      Doner Restaurant   

           9th Most Common Venue 10th Most Common Venue  
14                    Donut Shop              Drugstore  
23   Eastern European Restaurant      Electronics Store  
25                     Drugstore    Dumpling Restaurant  
30                     Drugstore    Dumpling Restaurant  
40                     Drugstore    Dumpling Restaurant  
44                    Donut Shop     

#### Cluster 3

In [40]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]
#manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 2, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Borough  Cluster Labels       1st Most Common Venue  \
91    Etobicoke             2.0  Construction & Landscaping   
97   North York             2.0              Baseball Field   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
91        Baseball Field         Women's Store     Electronics Store   
97         Women's Store     Electronics Store               Dog Run   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
91               Dog Run      Doner Restaurant            Donut Shop   
97      Doner Restaurant            Donut Shop             Drugstore   

   8th Most Common Venue        9th Most Common Venue  \
91             Drugstore          Dumpling Restaurant   
97   Dumpling Restaurant  Eastern European Restaurant   

         10th Most Common Venue  
91  Eastern European Restaurant  
97          Empanada Restaurant

#### Cluster 4

In [41]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]
#manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 3, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
19   North York             3.0                  Bank                  Café   
94    Etobicoke             3.0                  Bank         Women's Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
19   Japanese Restaurant    Chinese Restaurant         Women's Store   
94               Dog Run      Doner Restaurant            Donut Shop   

   6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
19      Doner Restaurant            Donut Shop                    Drugstore   
94             Drugstore   Dumpling Restaurant  Eastern European Restaurant   

   9th Most Common Venue       10th Most Common Venue  
19   Dumpling Restaurant  Eastern European Restaurant  
94     Electronics Store                        Diner

#### Cluster 5

In [42]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]
#manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 4, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
5   Scarborough             4.0            Playground         Women's Store   

         3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
5  Eastern European Restaurant        Discount Store               Dog Run   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
5      Doner Restaurant            Donut Shop             Drugstore   

  9th Most Common Venue 10th Most Common Venue  
5   Dumpling Restaurant      Electronics Store

### Thank you for completing this lab!

This notebook was created by [Alex Aklson](https://www.linkedin.com/in/aklson/) and [Polong Lin](https://www.linkedin.com/in/polonglin/). I hope you found this lab interesting and educational. Feel free to contact us if you have any questions!

This notebook is part of a course on **Coursera** called *Applied Data Science Capstone*. If you accessed this notebook outside the course, you can take this course online by clicking [here](http://cocl.us/DP0701EN_Coursera_Week3_LAB2).

<hr>

Copyright &copy; 2018 [Cognitive Class](https://cognitiveclass.ai/?utm_source=bducopyrightlink&utm_medium=dswb&utm_campaign=bdu). This notebook and its source code are released under the terms of the [MIT License](https://bigdatauniversity.com/mit-license/).